### PR TITLE
[llvm][Support] Stop using PWD in current_path

### PR DESCRIPTION
**Pull request by inaki-amatria**
In a filesystem structured as follows:

```
r/
|- a/
`- b -> a/
```

If current directory is set to `r/b` (a symbolic link to `r/a`) and then changed to `r/a` using `set_current_path()`, subsequent calls to `current_path()` incorrectly return `r/b`. This is because `current_path()` reads the `PWD` environment variable, which still points to the symbolic link `r/b` instead of the actual directory `r/a`.

This issue leads to inaccurate path reporting, causing confusion and potential errors in applications that rely on `current_path()`.

This issue is reproducible with bind mounts too.

Edit:

The reason for LLVM honoring `PWD` over `getcwd` is to align Clang with GCC, which allows changing paths via the `PWD` environment variable. This convention imposes unnecessary constraints on other users of the library. To address this, the `current_path` function has been modified to no longer rely on the `PWD` environment variable, ensuring accurate path resolution regardless of symbolic links or bind mounts.

### CODE REVIEW

**Comment by inaki-amatria** [comment]
CC @chandlerc

**Comment by inaki-amatria** [comment]
Ping

File: clang/test/Driver/coverage.c [hunk 1]
```diff
@@ -17,10 +17,6 @@
 // GCNO-LOCATION: "-coverage-notes-file={{.*}}/foo/bar.gcno"
 // GCNO-LOCATION-REL: "-coverage-notes-file={{.*}}{{/|\\\\}}foo/bar.gcno"
 
-/// GCC allows PWD to change the paths.
-// RUN: %if system-linux %{ PWD=/proc/self/cwd %clang -### -c --coverage %s -o foo/bar.o 2>&1 | FileCheck --check-prefix=PWD %s %}
-// PWD: "-coverage-notes-file=/proc/self/cwd/foo/bar.gcno" "-coverage-data-file=/proc/self/cwd/foo/bar.gcda"
```
Comment (AaronBallman) [inline clang/test/Driver/coverage.c:22]: We should not be regressing Clang behavior like this -- at least not without wider buy-in from the Clang community via an RFC. I worry that this change will silently break build scripts relying on the current PWD behavior being consistent between Clang and GCC.

Comment (inaki-amatria) [reply clang/test/Driver/coverage.c:22]: Thank you for the timely response and for raising this important concern.

I understand the importance of maintaining consistent behavior in Clang, and I certainly do not wish to introduce changes that might silently break existing build scripts.

Could you please guide me on how to initiate the RFC process? I would like to ensure that we address this issue appropriately.

Thank you again for your insights, and I look forward to your guidance on progressing with the RFC.

Comment (AaronBallman) [reply clang/test/Driver/coverage.c:22]: Sure! Basically, you'd write a post in https://discourse.llvm.org/ with a title along the lines of `[RFC]; stop using PWD in current_path` and some explanation of what you'd like to change, why, and what the impacts are. Folks will leave comments in the post about things you should also consider, and eventually discussion dies down and we (hopefully) have clear direction on what to do from there.

I would recommend posting this RFC in the [Clang category](https://discourse.llvm.org/c/clang/6) though it's possible the changes could impact other projects.

**Comment by inaki-amatria** [comment]
Closing this until I post in discourse. TYSM for your time, Aaron

**Pull request closed without merging**